server/routes/query_assist/utils/agents.ts

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { ApiResponse } from '@opensearch-project/opensearch/.';
+import { SearchResponse, SearchTotalHits } from '@opensearch-project/opensearch/api/types';
+import { RequestBody } from '@opensearch-project/opensearch/lib/Transport';
+import { OpenSearchClient } from '../../../../../../src/core/server';
+import { isResponseError } from '../../../../../../src/core/server/opensearch/client/errors';
+import { ML_COMMONS_API_PREFIX } from '../../../../common/constants/query_assist';
+
+const AGENT_REQUEST_OPTIONS = {
+  /**
+   * It is time-consuming for LLM to generate final answer
+   * Give it a large timeout window
+   */
+  requestTimeout: 5 * 60 * 1000,
+  /**
+   * Do not retry
+   */
+  maxRetries: 0,
+};
+
+type AgentResponse = ApiResponse<{
+  inference_results: Array<{
+    output: Array<{ name: string; result?: string }>;
+  }>;
+}>;
+
+export const agentIdMap: Record<string, string> = {};
+
+export const searchAgentIdByName = async (
+  opensearchClient: OpenSearchClient,
+  name: string
+): Promise<string> => {
+  try {
+    const response = (await opensearchClient.transport.request({
+      method: 'GET',
+      path: `${ML_COMMONS_API_PREFIX}/agents/_search`,
+      body: {
+        query: {
+          term: {
+            'name.keyword': name,
+          },
+        },
+        sort: {
+          created_time: 'desc',
+        },
+      },
+    })) as ApiResponse<SearchResponse>;
+
+    if (
+      !response ||
+      (typeof response.body.hits.total === 'number' && response.body.hits.total === 0) ||
+      (response.body.hits.total as SearchTotalHits).value === 0
+    ) {
+      throw new Error('cannot find any agent by name: ' + name);
+    }
+    const id = response.body.hits.hits[0]._id;
+    agentIdMap[name] = id;
```

Comment:
there is a global cache to avoid searching the agent id for every execute request. It should only search when the agent id is not cached or execute request failed with agent not found. is there a better way for this?

Reply:
Then I think in this case it's fine and should be safe. Other ways could be make it a class or have another function manages the step of fetching id and update cache, but we don't necessary have to.

Reply:
Then I think in this case it's fine and should be safe. Other ways could be make it a class or have another function manages the step of fetching id and update cache, but we don't necessary have to.

Reply:
make sense, will refactor to not do this inside search function